app/com/lynxanalytics/biggraph/frontend_operations/ImportOperations.scala

```diff
@@ -151,27 +145,85 @@ class ImportOperations(env: SparkFreeEnvironment) extends OperationRegistry {
+      ImportedDataParam(),
       new DummyParam("last_settings", ""))
 
-    def getRawDataFrame(context: spark.sql.SQLContext) = {
-      Neo4jUtil.read(
-        context,
-        params("node_label"),
-        params("relationship_type"),
-        splitParam("imported_columns").toSet,
-        params("infer") == "yes",
-        params("limit"),
-        params("num_partitions").toInt)
+    override def getOutputs(): Map[BoxOutput, BoxOutputState] = {
+      params.validate()
+      assert(params("imported_table").nonEmpty, "You have to import the data first.")
+      assert(!areSettingsStale, areSettingsStaleReplyMessage)
+      import CommonProjectState._
+      val state = json.Json.parse(params("imported_table")).as[CommonProjectState]
+      Map(context.box.output(context.meta.outputs(0)) -> BoxOutputState.from(state))
+    }
+
+    override def runImport(env: com.lynxanalytics.biggraph.BigGraphEnvironment): ImportResult = {
+      import org.apache.spark.sql.functions._
+      import org.apache.spark.sql.types
+      import com.lynxanalytics.biggraph.graph_util.Scripting._
+      val spark = SQLController.defaultContext()(env.sparkDomain)
+      val project = new RootProjectEditor(CommonProjectState.emptyState)
+      def runQuery(query: String, column: String) = {
+        import spark.implicits._
+        val json = spark.read.format("org.neo4j.spark.DataSource")
+          .option("authentication.type", "basic")
+          .option("authentication.basic.username", params("username"))
+          .option("authentication.basic.password", params("password"))
+          .option("url", params("url"))
+          .option("schema.strategy", "string")
+          .option("query", query)
+          .load
+          .persist(org.apache.spark.storage.StorageLevel.MEMORY_AND_DISK)
+        assert(json.columns.contains(column), s"You must set a query that returns '$column'.")
+        spark.read.json(json.select(col(column)).as[String])
+      }
+      // Get vertices.
+      if (params("vertex_query").nonEmpty) {
+        val df = runQuery(params("vertex_query"), "node")
+        val table = graph_operations.ImportDataFrame.run(df)
+        val vs = table.toProject
+        project.vertexSet = vs.vertexSet
+        project.vertexAttributes = vs.viewer.vertexAttributes
+      }
+      // Get edges.
+      if (params("edge_query").nonEmpty) {
+        val df = runQuery(params("edge_query"), "rel")
+        if (params("vertex_query").isEmpty) {
+          // Create vertices from the edge endpoints.
+          val vdf = df.select("`<source.id>`")
+            .union(df.select("`<target.id>`"))
+            .withColumnRenamed("<source.id>", "<id>")
+            .distinct
+          val table = graph_operations.ImportDataFrame.run(vdf)
+          val vs = table.toProject
+          project.vertexSet = vs.vertexSet
+          project.vertexAttributes = vs.viewer.vertexAttributes
```

Comment:
Why do we need to do this? Why not just read the edges and apply `Use table as graph` (well, more precisely `graph_operations.VerticesToEdges`)?

Reply:
> Why do we need to do this? Why not just read the edges and apply `Use table as graph` (well, more precisely `graph_operations.VerticesToEdges`)?

Done. It's not less code. Unfortunately we have to duplicate some code. The two operation outputs contain basically the same things (edges, embedding) but they are unrelated types, so we can't handle them together.

But still it should be faster. Thanks!